### PR TITLE
fix nodetool status to show zero-token nodes

### PR DESCRIPTION
**Pull request by AbhinavKumarJha**
In the current scenario, the nodetool status doesn’t display information regarding zero token nodes. For example, if 5 nodes are spun by the administrator, out of which, 2 nodes are zero token nodes, then nodetool status only shows information regarding the 3 non-zero token nodes.

This commit intends to fix this issue by leveraging the “/storage_service/host_id ” API  and adding appropriate logic in scylla-nodetool.cc to support zero token nodes.

A test is also added in nodetool/test_status.py to verify this logic. This test fails without this commit’s zero token node support logic, hence verifying the behavior.

This PR fixes a bug. Hence we need to backport it. Backporting needs to be done only
to 6.2 version, since earlier versions don't support zero token nodes.

Fixes: scylladb/scylladb#19849
Fixes: scylladb/scylladb#17857

### CODE REVIEW

File: locator/token_metadata.cc [hunk 1]
```diff
@@ -570,8 +570,7 @@ std::unordered_map<inet_address, host_id> token_metadata_impl::get_endpoint_to_h
     std::unordered_map<inet_address, host_id> map;
     map.reserve(nodes.size());
     for (const auto& [endpoint, node] : nodes) {
-        // Restrict to members
-        if (!node->is_member()) {
+        if (node->left() || node->is_none()) {
```
Comment (kbr-scylla) [inline locator/token_metadata.cc:573]: This change is not mentioned in the commit message. Does it fix some other bug?
Also, it probably belongs to a separate commit.

**Review by patjed41 — CHANGES_REQUESTED** [review]

File: test/nodetool/test_status.py [hunk 2]
```diff
@@ -510,3 +510,40 @@ def test_status_resolve(request, nodetool, resolve):
...
+        ),
+        Node(
+            endpoint="127.0.0.2",
+            host_id="ed341f60-b12a-4fd4-9917-e80977ded0f9",
+            load=277624,
+            tokens=[],
+            datacenter="datacenter1",
+            rack="rack2",
+            status=NodeStatus.Up,
+            state=NodeState.Joining,
```
[the remark this answers is not in the document]
Comment (kbr-scylla) [reply test/nodetool/test_status.py:535]: `Moving` state was removed a long time ago from Cassandra and I don't know if it was ever implemented in Scylla. Ignore it.

File: locator/token_metadata.cc [hunk 3]
```diff
@@ -570,8 +570,7 @@ std::unordered_map<inet_address, host_id> token_metadata_impl::get_endpoint_to_h
     std::unordered_map<inet_address, host_id> map;
```
Comment (kbr-scylla) [inline locator/token_metadata.cc:570]: I recall we discussed renaming `get_endpoint_to_host_id_map_for_reading`, did you abandon the idea?

File: locator/token_metadata.cc (hunk 1 above)
Comment (AbhinavKumarJha) [reply locator/token_metadata.cc]: Reverted it since we anyway don't display joining zero token nodes

File: locator/token_metadata.cc (hunk 3 above)
[the remark this answers is not in the document]
Comment (AbhinavKumarJha) [reply locator/token_metadata.cc]: Done

File: tools/scylla-nodetool.cc [hunk 4]
```diff
@@ -2125,14 +2135,13 @@ void status_operation(scylla_rest_client& client, const bpo::variables_map& vm)
...
-                    !is_effective_ownership_unknown ? format("{:.1f}%", endpoint_ownership.at(ep) * 100) : "?",
-                    endpoint_host_id.contains(ep) ? endpoint_host_id.at(ep) : "?",
-                    endpoint_rack.at(ep));
+            table.add(fmt::format("{}{}", status, state),
+                      address,
+                      load,
+                      token_count_unknown ? "?" : fmt::to_string(endpoint_tokens.contains(ep) ? endpoint_tokens.at(ep) : 0),
+                      !is_effective_ownership_unknown ? format("{:.1f}%", endpoint_ownership.at(ep) * 100) : "?",
+                      endpoint_host_id.contains(ep) ? endpoint_host_id.at(ep) : "?",
+                      endpoint_rack.at(ep));
```
[the remark this answers is not in the document]
Comment (AbhinavKumarJha) [reply tools/scylla-nodetool.cc]: Done

File: tools/scylla-nodetool.cc [hunk 5]
```diff
@@ -2091,6 +2091,16 @@ void status_operation(scylla_rest_client& client, const bpo::variables_map& vm)
...
+    for (const auto& [ep, host_id] : endpoint_host_id) {
+        if (endpoint_rack.contains(ep)) {
+            continue;
+        }
+        const auto dc = sstring(rjson::to_string_view(client.get("/snitch/datacenter", {{"host", ep}})));
+        const auto rack = sstring(rjson::to_string_view(client.get("/snitch/rack", {{"host", ep}})));
+        endpoint_rack.emplace(ep, rack);
+        dc_endpoints[dc].insert(ep);
+    }
+
```
[the remark this answers is not in the document]
Comment (AbhinavKumarJha) [reply tools/scylla-nodetool.cc]: > IIUC, all nodes appear in the new loop,

For this to be true, we need to always have host_id associated with nodes, which is sometimes not true for joining nodes. We have a test in test_status where host_id is unavailable. But this isn't a problem for zero token nodes since we anyway dont display joining nodes there. 

Hence a union of `tokens_endpoint` and `host_id_map` serves our purpose in all cases.

```
def test_status_keyspace_joining_node(request, nodetool):
    """ Joining nodes do not have some attributes available yet:
    * load
    * host_id
    """
```

File: test/nodetool/test_status.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (AbhinavKumarJha) [reply test/nodetool/test_status.py]: As discussed, we are not going forward with joining node display info. For leaving nodes, case is different, its not computed using tokens, it uses a getter for _leaving_endpoints data member, which is populated in `sync_raft_topology_nodes` . Hence it shows info regarding both regular nodes and zero token nodes

File: test/nodetool/test_status.py [hunk 6]
```diff
@@ -510,3 +510,40 @@ def test_status_resolve(request, nodetool, resolve):
     ]
 
     _do_test_status(request, nodetool, None, nodes, resolve)
+
+
+def test_status_with_zero_token_nodes(request, nodetool):
```
[the remark this answers is not in the document]
Comment (AbhinavKumarJha) [reply test/nodetool/test_status.py:515]: Done

File: locator/token_metadata.cc (hunk 3 above)
Comment (AbhinavKumarJha) [reply locator/token_metadata.cc]: Lets do that separately, since this is decoupled.

**Comment by AbhinavKumarJha** [comment]
Added topology tests for confirming behavior and removed all the zero token node joining display logic and tests.

File: test/nodetool/test_status.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (patjed41) [reply test/nodetool/test_status.py]: Right, leaving zero-token nodes are fine.

Now I remember that we call `token_metadata::add_leaving_endpoint` for zero-token nodes. The same goes for `token_metadata::add_replacing_endpoint`. It looks bad because logically zero-token nodes have nothing to do with the token ring. I decided to leave it like this when I worked on zero-token nodes because it's harmless and changing it would require more refactoring. The zero-token nodes PR was big enough without it.

I don't think we want to clean up this in the near future, but maybe we should at least open an issue to document it. @kbr-scylla

File: tools/scylla-nodetool.cc (hunk 5 above)
[the remark this answers is not in the document]
Comment (kbr-scylla) [reply tools/scylla-nodetool.cc]: > For this to be true, we need to always have host_id associated with nodes, which is sometimes not true for joining nodes. 

Which is a bug: https://github.com/scylladb/scylladb/issues/17857
I recall that we discussed it and I asked you to fix it as part of this PR, Abhinav.

> We have a test in test_status where host_id is unavailable.

So we're verifying that a bug is present? That doesn't make sense.

**Comment by kbr-scylla** [comment]
> This PR fixes a bug. Hence we need to backport it.

You applied backport/6.0 and backport/6.1 labels.
But zero-token nodes were introduced in 6.2, so it's unnecessary to backport to 6.0/6.1.

File: tools/scylla-nodetool.cc (hunk 5 above)
[the remark this answers is not in the document]
Comment (kbr-scylla) [reply tools/scylla-nodetool.cc]: The `/storage_service/host_id` api is calling `token_metadata::get_endpoint_to_host_id_map_for_reading` underneath.

This is how the function is implemented:
```cpp
std::unordered_map<inet_address, host_id> token_metadata_impl::get_endpoint_to_host_id_map_for_reading() const {
    const auto& nodes = _topology.get_nodes_by_endpoint();
    std::unordered_map<inet_address, host_id> map;
    map.reserve(nodes.size());
    for (const auto& [endpoint, node] : nodes) {
        // Restrict to members
        if (!node->is_member()) {
            continue;
        }
        if (const auto& host_id = node->host_id()) {
            map[endpoint] = host_id;
        } else {
            tlogger.info("get_endpoint_to_host_id_map_for_reading: endpoint {} has null host_id: state={}", endpoint, node->get_state());
        }
    }
    return map;
}
```

As we can see, it is skipping joining nodes:
```cpp
        // Restrict to members
        if (!node->is_member()) {
            continue;
        }
```

And perhaps that made sense at some point. The function name suffix `_for_reading` suggests that the function was used in some other places in the past, and indeed if we need endpoints "for reading" then we cannot show joining endpoints.

However, recall @AbhinavKumarJha that  we determined that this function is used **only** by this api endpoint today. So we can modify it as we like to suit our needs, and also change the name to better describe what it does after our fixes.


So if we modify the endpoint to include joining nodes, then joining nodes will appear in the `/storage_service/host_id` endpoint. Then, as @patjed41 observed, you should not need the first loop: all nodes will appear in the `for (const auto& [ep, host_id] : endpoint_host_id) {` loop. 

At the same time, this will fix https://github.com/scylladb/scylladb/issues/17857. Two birds with one stone!

Now my question is: in earlier version of this PR, I saw that you actually modified that `get_endpoint_to_host_id_map_for_reading` function. So it seems that you actually tried my proposal. So perhaps I'm missing something and it doesn't work?

[the remark this answers is not in the document]
Comment (AbhinavKumarJha) [reply tools/scylla-nodetool.cc]: >So we're verifying that a bug is present? That doesn't make sense

The test verifies that even if host_id is None, the nodetool will display the node. Since we already discussed that this case can never exist, I added `@pytest.mark.skip(reason="This scenario can never happen, all joining nodes deterministically have host_id available")` for that test.(Since we use host_id_map, this test fails with our implementation)

> At the same time, this will fix https://github.com/scylladb/scylladb/issues/17857. Two birds with one stone!

Sure, done

File: locator/token_metadata.cc (hunk 3 above)
Comment (AbhinavKumarJha) [reply locator/token_metadata.cc:570]: I made this name change!

File: locator/token_metadata.cc (hunk 1 above)
Comment (AbhinavKumarJha) [reply locator/token_metadata.cc:573]: Added it back as per our discussion on other comment.

>Also, it probably belongs to a separate commit.

We are directly using this logic in scylla-nodetool.cc, so I think we need it here in this commit. Lemme know if otherwise

File: tools/scylla-nodetool.cc (hunk 5 above)
[the remark this answers is not in the document]
Comment (kbr-scylla) [reply tools/scylla-nodetool.cc]: > The test verifies that even if host_id is None, the nodetool will display the node. Since we already discussed that this case can never exist, I added @pytest.mark.skip(reason="This scenario can never happen, all joining nodes deterministically have host_id available") for that test

We could keep it enabled, you could e.g. run newer nodetool against older version of Scylla, which still has https://github.com/scylladb/scylladb/issues/17857, and then the test verifies that in this case, nodetool will still show nodes with missing host IDs.

> Since we use host_id_map, this test fails with our implementation

It fails? Then I don't understand something:
- does the test run against mocked REST API endpoints? If it does, then it shouldn't matter what we do inside Scylla, the test should still pass because the mocked endpoint does not return host ID
- or does the test run against a real Scylla instance?

**Comment by AbhinavKumarJha** [comment]
> > This PR fixes a bug. Hence we need to backport it.
> 
> You applied backport/6.0 and backport/6.1 labels. But zero-token nodes were introduced in 6.2, so it's unnecessary to backport to 6.0/6.1.

After Patryk's comment, I removed those labels: 
image
Can you still see those?

**Comment by kbr-scylla** [comment]
>  Can you still see those?

They're gone now, thanks

File: locator/token_metadata.cc (hunk 1 above)
Comment (kbr-scylla) [reply locator/token_metadata.cc:573]: You can split the PR into two commits, this change (which fixes and renames `get_endpoint_to_host_id_map`) will go to the first one, as it fixes mostly unrelated bug and it makes sense as a standalone change. And deserves its own commit message. The second commit would contain the rest of the changes

File: test/topology/test_zero_token_node_nodetool.py [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+    exe_path = await manager.server_get_exe(servers[0].server_id)
+    cmd = [exe_path, "nodetool", "status"] + ["--logger-log-level",
+                                              "scylla-nodetool=trace",
+                                              "-h", servers[0].ip_addr]
+
+    result = subprocess.run(cmd, capture_output=True, text=True)
+
+    live_eps = await manager.api.client.get_json("/gossiper/endpoint/live", host=servers[0].ip_addr)
+    down_eps = await manager.api.client.get_json("/gossiper/endpoint/down", host=servers[0].ip_addr)
+    leaving = await manager.api.client.get_json("/storage_service/nodes/leaving", host=servers[0].ip_addr)
```
Comment (patjed41) [inline test/topology/test_zero_token_node_nodetool.py:72]: In https://github.com/scylladb/scylladb/pull/20909#discussion_r1782563199, I asked for a test that verifies API endpoints when zero-token are in the cluster. The test you added doesn't do this. For example, `live_eps + down_eps + leaving` could be empty and the test would pass.

What we should do here is test that `live_eps`, `down_eps`, `leaving` are correct. In this test case, `live_eps` should contain IPs of all nodes (including zero-token nodes), each IP should appear once, and only nodes' IPs should appear. Also, `down_eps` and `leaving` should be empty. Other test cases should work similarly, just with different conditions.

My initial idea was that we don't run `nodetool status` in this test and we only test all API endpoints used by it (`test_status_with_zero_token_nodes` already tests `nodetool status` assuming API endpoints work correctly). This would require us to also check other API endpoints like `/storage_service/host_id`.

But now I think that what you did can also be fine if you improve it. If you check in `validate_status_operation` that host IDs are correct, then we will know that `/storage_service/host_id` works correctly for zero-token nodes. You should also check that `nodetool status` shows 0 tokens for zero-token nodes and a correct rack.

I also like what you did because it allows us to test one more thing easily -- running `nodetool status` on a zero-token node. It should work just like on a token-owning node. You can run `nodetool status` twice in this test case (once on a zero-token node) and test both runs in the same way. In other test cases it may be not that easy, but you can try.

File: test/topology/test_zero_token_node_nodetool.py [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+
+    result = subprocess.run(cmd, capture_output=True, text=True)
+
+    post_wait_servers = await manager.running_servers()
+    live_eps = await manager.api.client.get_json("/gossiper/endpoint/live", host=servers[0].ip_addr)
+    down_eps = await manager.api.client.get_json("/gossiper/endpoint/down", host=servers[0].ip_addr)
+    leaving = await manager.api.client.get_json("/storage_service/nodes/leaving", host=servers[0].ip_addr)
+
+    await validate_status_operation(result.stdout, live_eps, down_eps, leaving)
+    [await manager.api.message_injection(s.ip_addr, 'delay_node_removal') for s in post_wait_servers]
```
Comment (patjed41) [inline test/topology/test_zero_token_node_nodetool.py:104]: I don't understand why you need `post_wait_servers`. The only difference could be that `post_wait_servers` contains the zero token node, and `servers` doesn't, but that's fine. There is no need to send the injection message to the zero-token node.

**Review by patjed41 — CHANGES_REQUESTED** [review]

File: service/topology_coordinator.cc [hunk 9]
```diff
@@ -1947,8 +1947,12 @@ class topology_coordinator : public endpoint_lifecycle_subscriber {
                                                    "bootstrap: read fence completed");
                     }
                     break;
-                case node_state::removing:
+                case node_state::removing: {
+                    rtlogger.info("Waiting for node removal");
```
[the remark this answers is not in the document]
Comment (patjed41) [reply service/topology_coordinator.cc:1951]: Also, I would change it to something like "delay_node_removal: waiting for message". We wait for a message there, not for node removal.

File: locator/token_metadata.cc (hunk 1 above)
Comment (AbhinavKumarJha) [reply locator/token_metadata.cc:573]: Done

File: tools/scylla-nodetool.cc (hunk 5 above)
[the remark this answers is not in the document]
Comment (AbhinavKumarJha) [reply tools/scylla-nodetool.cc]: As already discussed, I removed the test. Reasoning:

Tests are run against mock API endpoints only. For that particular test, host_id_map returned from API has not changed. What has changed is that earlier we used token_endpoint_map for discovering what nodes are present(lets call it v1). Now we are using endpoint_host_id map for that purpose(lets call it v2).
Now, the test that I am talking about has host_id=None which means that in v1, it would show the host_id as ?, but for v2 it won't display that node at all, since we are using endpoint_host_id map.

File: service/topology_coordinator.cc (hunk 9 above)
[the remark this answers is not in the document]
Comment (AbhinavKumarJha) [reply service/topology_coordinator.cc]: Done

File: test/topology/test_zero_token_node_nodetool.py [hunk 10]
```diff
@@ -0,0 +1,128 @@
+import subprocess
+import asyncio
+from collections.abc import Coroutine
+from test.pylib.manager_client import ManagerClient
+import pytest
+
+
+@pytest.mark.asyncio
```
[the remark this answers is not in the document]
Comment (AbhinavKumarJha) [reply test/topology/test_zero_token_node_nodetool.py]: Done, are we removing here because 'await' has not been used?

File: test/topology/test_zero_token_node_nodetool.py [hunk 11]
```diff
@@ -0,0 +1,128 @@
...
+
+        if ep in live_eps:
+            assert status_state[0] == 'U'
+        else:
+            assert status_state[0] == 'D'
+
+        if ep in leaving:
+            assert status_state[1] == 'L'
+        else:
+            assert status_state[1] == 'N'
```
[the remark this answers is not in the document]
Comment (AbhinavKumarJha) [reply test/topology/test_zero_token_node_nodetool.py]: Done

[the remark this answers is not in the document]
Comment (AbhinavKumarJha) [reply test/topology/test_zero_token_node_nodetool.py]: Done

File: test/topology/test_zero_token_node_nodetool.py [hunk 12]
```diff
@@ -0,0 +1,128 @@
...
+
+
+async def wait_for_first_completed(coros: list[Coroutine]):
+    done, pending = await asyncio.wait([asyncio.create_task(c) for c in coros], return_when=asyncio.FIRST_COMPLETED)
+    for t in pending:
+        t.cancel()
+    for t in done:
+        await t
```
[the remark this answers is not in the document]
Comment (AbhinavKumarJha) [reply test/topology/test_zero_token_node_nodetool.py]: Done

File: test/topology/test_zero_token_node_nodetool.py [hunk 13]
```diff
@@ -0,0 +1,128 @@
...
+
+
+@pytest.mark.asyncio
+async def test_zero_token_node_normal(manager: ManagerClient):
+    await manager.servers_add(3)
```
[the remark this answers is not in the document]
Comment (AbhinavKumarJha) [reply test/topology/test_zero_token_node_nodetool.py]: Done.

>Alternatively, you can move the test to topology_custom

What's the difference and when do we use custom/

File: test/topology/test_zero_token_node_nodetool.py (hunk 7 above)
Comment (AbhinavKumarJha) [reply test/topology/test_zero_token_node_nodetool.py]: Done.

>In other test cases it may be not that easy, but you can try.
Not possible for other cases, since node is down

File: test/topology/test_zero_token_node_nodetool.py (hunk 8 above)
Comment (AbhinavKumarJha) [reply test/topology/test_zero_token_node_nodetool.py]: Done

File: test/topology/test_zero_token_node_nodetool.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (kbr-scylla) [reply test/topology/test_zero_token_node_nodetool.py]: topology_custom has `initial_size: 0`. E.g. when you want to boot a cluster with non-standard configuration you need to use topology_custom

File: test/topology/test_zero_token_node_nodetool.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (patjed41) [reply test/topology/test_zero_token_node_nodetool.py]: It looks like `@pytest.mark.asyncio` does something only if it is used with a test function (named `test_*`): https://pytest-asyncio.readthedocs.io/en/latest/reference/markers/index.html

**Comment by kbr-scylla** [comment]
@AbhinavKumarJha you pushed to the branch multiple times, but you didn't post a summary comment of what changed in the new version.
Please post the summary comment.

**Comment by AbhinavKumarJha** [comment]
>Added topology tests for confirming behavior and removed all the zero token node joining display logic and tests.

Not much after this summary, disabled the tests for release mode, since they were failing due to lack of error injection support, also added a confirmation logic for joining node to be in "UJ" state

File: api/token_metadata.cc [hunk 14]
```diff
@@ -71,7 +71,7 @@ void set_token_metadata(http_context& ctx, routes& r, sharded<locator::shared_to
```
Comment (kbr-scylla) [inline api/token_metadata.cc:71]: Regarding the commit `token_metadata: rename endpoint_to_host_id_map getter ...`

If I understand correctly, this commit is fixing https://github.com/scylladb/scylladb/issues/17857
please confirm -- and if it's true, then add `Fixes: scylladb/scylladb#17857` at the end of the commit message. And also at the end of the PR description.

File: test/pylib/util.py [hunk 15]
```diff
@@ -246,3 +247,11 @@ async def view_is_built():
         return done[0][0] == node_count or None
     deadline = time.time() + timeout
     await wait_for(view_is_built, deadline)
+
+
+async def wait_for_first_completed(coros: list[Coroutine]):
```
Comment (kbr-scylla) [inline test/pylib/util.py:252]: If I understand correctly, the idea behind Patryk's previous comment was that such function already exists in existing test: `test_coordinator_queue_management.py`

You don't want to duplicate the code. Instead, you want to **move** the code.

So in addition to adding the function here in test/pylib/util.py, you should also remove it from test_coordinator_queue_management.py and instead `import` it from in that test.

File: test/topology/test_nodetool.py [hunk 16]
```diff
@@ -0,0 +1,208 @@
...
+            assert status_state[1] == 'J'
+        elif ep in leaving:
+            assert status_state[1] == 'L'
+        else:
+            assert status_state[1] == 'N'
+
+        if ep in zero_token_eps:
+            assert int(tokens) == 0
+        else:
+            assert int(tokens) != 0
```
Comment (kbr-scylla) [inline test/topology/test_nodetool.py:59]: You can (and probably should) check the exact value.
It is defined in `make_scylla_conf` in test/pylib/scylla_cluster.py:
```python3
        'num_tokens': 16,
```

You could extract the `16` into a constant in scylla_cluster.py, and import the constant here, and verify that `int(tokens)` is equal to it.

File: test/topology/test_nodetool.py [hunk 17]
```diff
@@ -0,0 +1,208 @@
...
+
+        assert host_id != "" and host_id != "?"
```
Comment (kbr-scylla) [inline test/topology/test_nodetool.py:61]: You can (and almost definitely should) check for the exact value.

It's easy to obtain the host IDs of our servers in each test (`ManagerClient.get_host_id`). Then you could for example turn `live_eps`, `down_eps` and so on from sets of IP addresses to maps `IP -> host ID` to pass the host ID of each node to this function.

File: test/topology/test_nodetool.py [hunk 18]
```diff
@@ -0,0 +1,208 @@
+import pytest
+import asyncio
+import subprocess
+from test.pylib.manager_client import ManagerClient
+from test.pylib.util import wait_for_first_completed
+from test.topology.conftest import skip_mode
+
+
+async def validate_status_operation(result, live_eps, down_eps, leaving, joining, zero_token_eps):
```
Comment (kbr-scylla) [inline test/topology/test_nodetool.py:9]: How about some type hints?
And use `mypy` to verify them

**Review by kbr-scylla — CHANGES_REQUESTED** [review]

File: test/topology/test_nodetool.py (hunk 17 above)
Comment (AbhinavKumarJha) [reply test/topology/test_nodetool.py]: Done

**Review by kbr-scylla — APPROVED** [review]

File: test/pylib/util.py [hunk 19]
```diff
@@ -3,6 +3,7 @@
 #
```
Comment (patjed41) [inline test/pylib/util.py:3]: The title of this commit message is very long. It could be something like "test: move `wait_for_first_completed` to `pylib/util.py`".

You should always try to make titles as short as possible. The only information included should be what you do and where without any details. "which returns on completion of any of the supplied coroutines" is a detail. If it was important, you should mention it below the title in a separate paragraph. But in this case it's not important -- the name of the function is intuitive and the diff already shows what it does.

What is important, however, is **why** you do this change. You should add the explanation below the title: you want to reuse this function in a new test added in the following commit and you want to avoid code duplication.

I think this commit is an important case so I'll elaborate. Writing a commit message like "I do this here because I do that in one of the following commits" may feel a bit weird initially, but it's actually a common pattern. You often want to do a big change (like adding a new test), but you need to do some self-contained adjustments first (like changing the testing framework). Then, you do these adjustments in separate commits before the big change to make the PR more readable -- you don't want one huge commit with many different changes and you can't do the adjustments after the big change because it depends on the adjustments. So, it's the only thing you can do. However, note that a PR structured this way can be very confusing for reviewers because they usually review PRs commit by commit. If for some commit there is no information that one of the following commits is relevant, it's pretty bad.

File: test/nodetool/test_status.py [hunk 20]
```diff
@@ -462,11 +462,8 @@ def test_status_keyspace_multi_dc(request, nodetool, uses_tablets, table):
     _do_test_status(request, nodetool, status_target, nodes)
 
 
-def test_status_keyspace_joining_node(request, nodetool):
-    """ Joining nodes do not have some attributes available yet:
-    * load
-    * host_id
-    """
```
Comment (patjed41) [inline test/nodetool/test_status.py:469]: Explain why you remove `test_status_keyspace_joining_node` in the commit message -- it's not obvious. I see the explanation here: https://github.com/scylladb/scylladb/pull/20909#discussion_r1808428713. The rest of this commit message is good.

Also, the diff of `test_status.py` unfortunately doesn't look like it should. You removed one test and and added another. But instead of one red "-" block and one green "+" block, there is a hard to decipher mess. Maybe a simple trick like changing host ID of the first node in `test_status_with_zero_token_nodes` (so it differs from the node in `test_status_resolve`) would fix it. You can try. I don't think fixing it is necessary, though. It's not like you did something wrong.

**Review by patjed41 — COMMENTED** [review]
All the changes look fine, good job.

I only left some comments regarding the commit messages. I would be happy if you made the changes I requested, but I don't think they are important and we can just merge the current version of this PR @kbr-scylla. Anyway, at least read what I wrote and think about it.

File: api/token_metadata.cc (hunk 14 above)
[the remark this answers is not in the document]
Comment (AbhinavKumarJha) [reply api/token_metadata.cc:71]: Done, added explanation!

File: test/pylib/util.py (hunk 19 above)
Comment (AbhinavKumarJha) [reply test/pylib/util.py:3]: Done, added explanation!

File: test/nodetool/test_status.py (hunk 20 above)
Comment (AbhinavKumarJha) [reply test/nodetool/test_status.py:469]: Done, added explanation!

>Maybe a simple trick like changing host ID of the first node in test_status_with_zero_token_nodes (so it differs from the node in test_status_resolve) would fix it

Had tried a few of these tricks, what happens is that the diff remains as it is right now, and just the new host_id string is green and the older one is made red, hence further increasing the chaos!

**Review by patjed41 — APPROVED** [review]
